ada/language/ast.py

```diff
@@ -3293,10 +3306,24 @@ def next_part():
 
         return Entity.match(
             lambda itd=T.IncompleteTypeDecl:
+            # The next part of a (non-private) incomplete type declaration must
+            # either be in the same declarative scope...
             itd.node_env
             .get(itd.name.name_symbol, LK.flat, categories=noprims)
             .find(lambda t: t.is_a(BaseTypeDecl) & (t != Entity))
-            .cast(BaseTypeDecl),
+
+            # Or in the particular case where the incomplete decl is in the
+            # private part of the package spec, the next part can be found in
+            # the package's body (RM 3.10.1).
+            ._or(Entity.is_in_private_part.then(
+                lambda _:
+                Entity.parent.parent.parent
```

Comment:
`.parent.parent.parent` seriously :disappointed: . Try `.declarative_scope` instead.

Reply:
Okay I changed this part as well as the existing piece of code below.